Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -925,28 +925,116 @@ extension SwiftLanguageServer {
     }
   }
 
+  func retrieveRefactorCodeActions(_ params: CodeActionRequest, completion: @escaping CodeActionProviderCompletion) {
+    guard let snapshot = documentManager.latestSnapshot(params.textDocument.url) else {
+      log("failed to find snapshot for url \(params.textDocument.url)")
+      completion([])
+      return
+    }
+    guard let startOffset = snapshot.utf8Offset(of: params.range.lowerBound),
+          let endOffset = snapshot.utf8Offset(of: params.range.upperBound) else
+    {
+        completion([])
+        return
+    }
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.cursorinfo
+    skreq[keys.sourcefile] = snapshot.document.url.path
+    skreq[keys.offset] = startOffset
+    let length = endOffset - startOffset
+    skreq[keys.length] = length
+    skreq[keys.retrieve_refactor_actions] = 1
+    if let settings = buildSystem.settings(for: snapshot.document.url, snapshot.document.language) {
+      skreq[keys.compilerargs] = settings.compilerArguments
+    }
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else {
+        completion([])
```

Comment:
This doesn't need `weak self` since sourcekitd will not store it after it calls it once (and it always calls it, even if the server crashes or the request is cancelled).